components/script/dom/path2d.rs

```diff
@@ -0,0 +1,315 @@
+}
+
+impl Path2D {
+    pub(crate) fn new() -> Path2D {
+        Self {
+            reflector_: Reflector::new(),
+            path: RefCell::new(vec![]),
+        }
+    }
+    pub(crate) fn new_with_path(other: &Path2D) -> Path2D {
+        Self {
+            reflector_: Reflector::new(),
+            path: other.path.clone(),
+        }
+    }
+    pub(crate) fn new_with_str(path: &str) -> Path2D {
+        let mut path_segments = Vec::new();
+        for result in SimplifyingPathParser::from(path) {
+            let Ok(seg) = result else { break };
+            let seg = match seg {
+                svgtypes::SimplePathSegment::MoveTo { x, y } => PathSegment::MoveTo {
+                    x: x as f32,
+                    y: y as f32,
+                },
+                svgtypes::SimplePathSegment::LineTo { x, y } => PathSegment::LineTo {
+                    x: x as f32,
+                    y: y as f32,
+                },
+                svgtypes::SimplePathSegment::CurveTo {
+                    x1,
+                    y1,
+                    x2,
+                    y2,
+                    x,
+                    y,
+                } => PathSegment::Bezier {
+                    cp1x: x1 as f32,
+                    cp1y: y1 as f32,
+                    cp2x: x2 as f32,
+                    cp2y: y2 as f32,
+                    x: x as f32,
+                    y: y as f32,
+                },
+                svgtypes::SimplePathSegment::Quadratic { x1, y1, x, y } => PathSegment::Quadratic {
+                    cpx: x1 as f32,
+                    cpy: y1 as f32,
+                    x: x as f32,
+                    y: y as f32,
+                },
+                svgtypes::SimplePathSegment::ClosePath => PathSegment::ClosePath,
+            };
+            path_segments.push(seg);
+        }
+        Self {
+            reflector_: Reflector::new(),
+            path: RefCell::new(path_segments.into_iter().map(PathSegmentWrapper).collect()),
+        }
+    }
+    pub(crate) fn push(&self, seg: PathSegment) {
+        self.path.borrow_mut().push(PathSegmentWrapper(seg));
+    }
+    pub(crate) fn segments(&self) -> Vec<PathSegment> {
+        self.path
+            .borrow()
+            .clone()
+            .into_iter()
+            .map(|PathSegmentWrapper(seg)| seg)
+            .collect()
+    }
+}
+
+impl Path2DMethods<crate::DomTypeHolder> for Path2D {
+    // https://html.spec.whatwg.org/multipage/#dom-path2d-addpath
+    fn AddPath(&self, other: &Path2D) {
+        let mut dest = self.path.borrow_mut();
+        dest.extend(other.path.borrow().iter().copied());
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-context-2d-closepath
+    fn ClosePath(&self) {
+        self.push(PathSegment::ClosePath);
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-context-2d-moveto
+    fn MoveTo(&self, x: f64, y: f64) {
+        if !([x, y].iter().all(|x| x.is_finite())) {
```

Comment:
I think this would be clearer when written as
```rust
x.is_finite() && y.is_finite()
```
and the same is true for the other methods that have this check